### ISSUE TITLE
Admin API via PulsarProxy with authentication fails

### ISSUE THREAD

**Issue by nkurihar** [apache/pulsar#2749]
I tried Admin API via PulsarProxy with Athenz authentication, and found a few unexpected behaviors. Please correct me if my expectation/understanding is wrong.

##### Expected
[`proxyRoles`](https://github.com/apache/pulsar/blob/90aa408b9ddc53629bdc5c0e540cf5335c34399b/conf/broker.conf#L206-L208) in broker.conf can be set to empty. 
(In this case Proxy will be treated as a "normal" client)

##### Actual
`proxyRoles` in broker.conf can NOT be set to empty.

Because:
[`validateOriginalPrincipal`](https://github.com/apache/pulsar/blob/90aa408b9ddc53629bdc5c0e540cf5335c34399b/pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java#L158-L162) will return false when
* request has `originalPrincipal`
* `proxyRoles` is empty

##### Proposal
Skip `validateOriginalPrincipal` if `proxyRoles` is empty.

##### Expected
A role can be both `proxyRoles` and [`superUserRoles`](https://github.com/apache/pulsar/blob/90aa408b9ddc53629bdc5c0e540cf5335c34399b/conf/broker.conf#L260-L262).

##### Actual
A role can be both `proxyRoles` and `superUserRoles`.
However, when we use this role as a "normal" client, its super user access to Admin API will fail.

Because:
Although the role behaves as a normal client, the process will enter [this if statement](https://github.com/apache/pulsar/blob/69071355ca03fac9e38f7501b37ca93ac19fef7b/pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java#L182) (since the role is also in `proxyRoles`) and finally [throw an exception](https://github.com/apache/pulsar/blob/69071355ca03fac9e38f7501b37ca93ac19fef7b/pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java#L187-L191) since [`originalPrincipalAuthorized`](https://github.com/apache/pulsar/blob/69071355ca03fac9e38f7501b37ca93ac19fef7b/pulsar-broker/src/main/java/org/apache/pulsar/broker/web/PulsarWebResource.java#L185) is false. (here `originalPrincipal` is null)

##### Proposal
If `originalPrincipal` is null, treat a role as a "normal" client even when it is in `proxyRoles`.

##### Expected
HTTP header includes authentication data of proxy.

##### Actual
HTTP header includes authentication data of original client.

Example of HTTP header from proxy to broker:
```
User-Agent: Jersey/2.23.2 (HttpUrlConnection 1.8.0_144)
Host: <broker>:<broker-port>
Accept: application/json
Athenz-Role-Auth: <token of original client> # This should be token of proxy.
Via: http/1.1 <proxy>
X-Forwarded-For: <client-ip>
X-Forwarded-Proto: http
X-Forwarded-Host: <proxy>
X-Forwarded-Server: <proxy-ip>
X-Original-Principal: <role of original client>
```

##### Proposal
Move original client's authentication data to another new header (e.g. X-Original-Authentication), and overwrite the authentication header with  proxy's authentication data.

#### System configuration
* Pulsar version: 2.1.1
* broker.conf: `authenticateOriginalAuthData=false`
* proxy.conf: `forwardAuthorizationCredentials=false`

**Comment by sijie** [comment]
@ivankelly can you help with this?

**Comment by nkurihar** [comment]
Hi, I added my proposals to fix this issue.
Please check whether it is reasonable or not.
Then I will create PR.

**Comment by ivankelly** [comment]
> Skip validateOriginalPrincipal if proxyRoles is empty.

This could open up a security hole, as if the proxyRoles is empty, any client could then pass any value as the validateOriginalPrincipal, a therefore escalate their privileges.

> If originalPrincipal is null, treat a role as a "normal" client even when it is in proxyRoles.

The point of proxyRoles is to limit which roles can pretend to be other roles. Allowing proxyRoles to be ignored defeats this purpose.

The proxyRoles should be dedicated roles that are used for nothing but the proxy. I would consider anything else a bad configuration. Maybe we could improve the logging to make it clearer to the administrator that proxyRoles needs to be configured.

**Comment by nkurihar** [comment]
Thanks your comments.

In producer/consumer proxied requests(not Admin REST API), we can use proxy without setting `proxyRoles` for now.
I have understood that as default behavior: 
https://pulsar.apache.org/docs/en/security-authorization/#proxy-roles
> By default, the broker treats the connection between a proxy and the broker as a normal user connection.

* Should `proxyRoles` be mandatory when we use proxy?
* Is a role in `proxyRoles` not allowed to behave as a "normal" client?

**Comment by ivankelly** [comment]
> Should proxyRoles be mandatory when we use proxy?

I think yes, as we give the proxy special abilities to spoof identity and this is a very powerful ability. However, my thinking on this is 100% TLS auth focussed, so I don't know how this interacts with athenz. Also, if auth is completely off between the proxy and broker, this would be unnecessary, though I don't know if this is currently the case.

But it would be good to get some other opinions. @merlimat @jai1 @joefk @massakam @sijie 

> Is a role in proxyRoles not allowed to behave as a "normal" client?

Also yes, modulo other opinions.

**Comment by nkurihar** [comment]
In my opinion, regardless of auth methods(TLS, Athenz, etc.), authentication/authorization of original client in broker side is not necessary but optional since proxy has already done it.

**Comment by ivankelly** [comment]
@nkurihar the issue is that proxies are exposed to the world, so more likely to be compromised. If the proxy is compromised at that point that proxy has full control of the system. With TLS and without proxyRoles, there's no way to block this access without regenerating all certs. With TLS and proxyRoles, you just need to reconfigure proxyRoles to exclude the role of the proxy that was compromised.
(thinking out loud here) But in this case, the proxy would fall back to being a normal role, and since the proxy role needs access to all the resources, it would still have access to the same resources.

Again, other opinions would be very useful here, especially from the people who originally added proxy roles. @merlimat @jai1 @joefk @massakam @sijie

**Comment by sijie** [comment]
@sijie @merlimat please help review the comments.

**Comment by sijie** [comment]
@sijie @merlimat please help review the comments.

**Comment by jai1** [comment]
The logic in validateOriginalPrincipal in PulsarWebResource seems incorrect it should loosely match the logic in invalidOriginalPrincipal in ServerCnx.java. 

I will raise a PR with the fix by tomorrow EOD.

Yes - proxyRoles can be empty but it's not recommended.

**Comment by jai1** [comment]
@nkurihar - Can you help review and approve #2986

**Comment by sijie** [comment]
This is fixed by #2986
